### DISCUSSION TITLE
How can I create a custom deserializer that warns when items aren't found?

### DISCUSSION THREAD

**Discussion by MHebes** [nlohmann/json#2684] [Q&A]
I'm trying to create a custom (de)serializer, following the guidance on [this page](https://nlohmann.github.io/json/features/arbitrary_types/#can-i-write-my-own-serializer-advanced-use) and [this test suite](https://github.com/nlohmann/json/blob/develop/test/src/unit-udt.cpp).

My desired behavior for deserialization to a struct is that if a key doesn't exist, or a key does exist and the type is wrong, or the array index is out of range or something, that element is just ignored and e.g. a warning is printed.

I basically want the deserializer to keep going if it hits an exception.

I have tried this so far:

```c++
// NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE but templated to support any basic_json
#define JSON_SERIALIZE(Type, ...)                                              \
  template <typename BasicJsonType>                                            \
  void to_json(BasicJsonType &nlohmann_json_j, const Type &nlohmann_json_t) {  \
    NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_TO, __VA_ARGS__))   \
  }                                                                            \
  template <typename BasicJsonType>                                            \
  void from_json(const BasicJsonType &nlohmann_json_j,                         \
                 Type &nlohmann_json_t) {                                      \
    NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_FROM, __VA_ARGS__)) \
  }

namespace Settings {

template <typename T, typename = void>
struct settings_serializer;

// non-defaults are ordered_map and settings_serializer
using json =
    nlohmann::basic_json<nlohmann::ordered_map, std::vector, std::string, bool,
                         std::int64_t, std::uint64_t, double, std::allocator,
                         settings_serializer>;

template <typename T, typename>
struct settings_serializer {
  static void from_json(const Settings::json &j, T &t) {
    std::cout << "settings_serializer::from_json for " << j
              << std::endl;
    // just default adl
    using nlohmann::from_json;
    try {
      from_json(j, t);
    } catch (std::exception &e) {
      std::cout << "ERROR: " << e.what() << std::endl;
    }
  }

  static void to_json(Settings::json &j, const T &t) {
    // just default adl
    using nlohmann::to_json;
    to_json(j, t);
  }
};

}  // namespace Settings
```

With the usage:

```c++
struct MySettings {
  bool SomeBool = false;
  double SomeDouble = 123.4;
};
JSON_SERIALIZE(MySettings, SomeBool, SomeDouble)

int main(int argc, char **argv) {
  using namespace std;

  // invokes to_json four times
  Settings::json config = {
      {"SomeBool", true},
      {"SomeDouble", 1.0},
  };

  // invokes from_json  3 times
  MySettings m = config;

  // invokes to_json 3 times
  Settings::json j = m;

  cout << j.dump(4) << endl;
}
```

Which prints:

```
default settings_serializer::from_json for {"SomeBool":true,"SomeDouble":1.0}
default settings_serializer::from_json for true
default settings_serializer::from_json for 1.0
{
    "SomeBool": true,
    "SomeDouble": 1.0
}
```

As expected.

However, if I change `Settings::json config` to be:

```c++
  Settings::json config = {
      // {"SomeBool", true},
      {"SomeDouble", 1.0},
  };
```

It chokes on SomeBool and quits deserializing:

```
settings_serializer::from_json for {"SomeDouble":1.0}
ERROR: [json.exception.out_of_range.403] key 'SomeBool' not found
{
    "SomeBool": false,
    "SomeDouble": 123.4
}
```

Instead, I would like it to print the error, but continue with the next sibling in the tree, so that `SomeBool` stays uninitialized (`false`) but `SomeDouble` is still set to `1.0`.

Is there any way I can do this?

I also attempted messing around with changing `to_json` and `from_json` directly (and the associated macros), but that's a bad solution for me because I don't want to override the normal `nlohmann::json` parsing for `MySettings` so that I have the option to do things like:

```c++
MySettings s;
Settings::json customJson = s;
nlohmann::json normalJson = customJson;
```

Any advice? How can I prevent the deserializer from stopping when it hits an exception?

**Comment by nlohmann** [comment]
There is no built-in 'continue on error' mode for the NLOHMANN_DEFINE_TYPE_* macros or for from_json in general - the generated code (NLOHMANN_JSON_FROM) calls `.at(key).get_to(member)` for every member in sequence, and `.at()` throws immediately on a missing/mismatched key, aborting the whole from_json call. To get per-field resilience you cannot rely on the macro; you need to hand-write from_json and wrap each member access individually, e.g. `try { j.at("SomeBool").get_to(t.SomeBool); } catch (const json::exception& e) { std::cerr << e.what(); }` for every field (or use `j.value("SomeBool", t.SomeBool)` if you're fine with silent defaulting instead of a warning). This can still live in a generic adl_serializer/settings_serializer so ordinary code keeps calling get<T>()/assignment as usual; it just means giving up the macro-generated body for this particular type.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
